File: examples/bench_10gb.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT OR Apache-2.0
#
# 10-iteration per-object benchmark at 10 GB.
#
# Each iteration:
#   1. Print `free -h` to show available memory before generation.
#   2. Generate 10 GB via Generator(size=N).get_chunk(N) — new Generator
#      (and DataGenerator + Rayon thread pool) per call.  This is EXACTLY
#      how dlio_benchmark calls dgen-py for large tensors.
#   3. Record elapsed time.
#   4. Print the result.
#   5. Explicitly del the BytesView + gc.collect() to release the 10 GB buffer.
#   6. Print `free -h` to show memory returned to OS.
#
# Usage:
#   source .venv/bin/activate
#   python examples/bench_10gb.py

import gc
import subprocess
import sys
import time

try:
    import dgen_py
    from dgen_py import Generator
except ImportError:
    print("ERROR: dgen-py not installed.  Run:  source .venv/bin/activate")
    sys.exit(1)

SIZE = 10 * 1024 ** 3  # 10 GB
RUNS = 10


def free():
    result = subprocess.run(["free", "-h"], capture_output=True, text=True)
    print(result.stdout, end="")


def main():
    print(f"=== dgen-py 10 GB per-object benchmark  ({RUNS} runs) ===")
    print("  Calls Generator(size=10 GB).get_chunk(10 GB) — new Generator every call.")
    print("  Buffer explicitly del'd + gc.collect() after each timed call.")
    print()

    times = []

    for i in range(1, RUNS + 1):
        print(f"── Run {i}/{RUNS}  (before generation) ──")
        free()

        t0 = time.perf_counter()
        gen = Generator(size=SIZE)
        bv = gen.get_chunk(SIZE)
        elapsed = time.perf_counter() - t0

        actual_bytes = len(bv) if bv is not None else 0
        gb_s = actual_bytes / elapsed / 1e9
        print(f"  → {elapsed:.3f} s   {gb_s:.2f} GB/s   ({actual_bytes} bytes)")
        times.append(gb_s)

        # Explicitly release the 10 GB buffer before the next iteration.
        del bv
        del gen
        gc.collect()

        print(f"── Run {i}/{RUNS}  (after free) ──")
        free()
        print()

    avg = sum(times) / len(times)
    mn  = min(times)
    mx  = max(times)
    srt = sorted(times)
    med = srt[len(srt) // 2]

    print("=== Summary ===")
    print(f"  Runs:    {RUNS}")
    for idx, v in enumerate(times, 1):
        print(f"  Run {idx:2d}:  {v:.2f} GB/s")
    print()
    print(f"  Min:     {mn:.2f} GB/s")
    print(f"  Median:  {med:.2f} GB/s")
    print(f"  Avg:     {avg:.2f} GB/s")
    print(f"  Max:     {mx:.2f} GB/s")


if __name__ == "__main__":
    main()

File: examples/bench_10gb.rs
// examples/bench_10gb.rs
//
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// 10-iteration per-object benchmark at 10 GB.
//
// Each iteration:
//   1. Print `free -h` to show available memory before generation.
//   2. Generate 10 GB via generate_data_simple() (new Rayon pool per call,
//      matching the large-object path in sai3-bench's data_gen_pool.rs).
//   3. Record elapsed time.
//   4. Print the result.
//   5. Explicitly drop the buffer with drop().
//   6. Print `free -h` to show memory returned to OS.
//
// Usage:
//   cargo run --release --example bench-10gb

use dgen_data::generate_data_simple;
use std::process::Command;
use std::time::Instant;

const SIZE: usize = 10 * 1024 * 1024 * 1024; // 10 GB
const RUNS: usize = 10;

fn free() {
    let out = Command::new("free")
        .arg("-h")
        .output()
        .expect("failed to run free");
    print!("{}", String::from_utf8_lossy(&out.stdout));
}

fn main() {
    println!("=== 10 GB per-object benchmark  ({RUNS} runs) ===");
    println!("  Calls generate_data_simple(10 GB) — new Rayon pool every call.");
    println!("  Buffer explicitly drop()'d after each timed call.");
    println!();

    let mut times = Vec::with_capacity(RUNS);

    for i in 1..=RUNS {
        println!("── Run {i}/{RUNS}  (before generation) ──");
        free();

        let t0 = Instant::now();
        let buf = generate_data_simple(SIZE, 1, 1);
        let elapsed = t0.elapsed().as_secs_f64();
        let gb_s = buf.len() as f64 / elapsed / 1e9;

        println!(
            "  → {:.3} s   {:.2} GB/s   ({} bytes)",
            elapsed,
            gb_s,
            buf.len()
        );
        times.push(gb_s);

        // Explicitly free the 10 GB buffer before the next iteration.
        drop(buf);

        println!("── Run {i}/{RUNS}  (after drop) ──");
        free();
        println!();
    }

    let avg = times.iter().sum::<f64>() / times.len() as f64;
    let min = times.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = times.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    // median
    let mut sorted = times.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let median = sorted[sorted.len() / 2];

    println!("=== Summary ===");
    println!("  Runs:    {RUNS}");
    for (i, v) in times.iter().enumerate() {
        println!("  Run {:2}:  {:.2} GB/s", i + 1, v);
    }
    println!();
    println!("  Min:     {min:.2} GB/s");
    println!("  Median:  {median:.2} GB/s");
    println!("  Avg:     {avg:.2} GB/s");
    println!("  Max:     {max:.2} GB/s");
}
